### PR TITLE
Don't downgrade on prerelease `VersionReq` when update with --breaking.

### PR DESCRIPTION
**Pull request by linyihai**
### What does this PR try to resolve?
Do nothing with prerelease when update with `--breaking`. 

Fixes https://github.com/rust-lang/cargo/issues/14178

### How should we test and review this PR?

Previous commit add a test, next commit to fix and update.

### CODE REVIEW

File: src/cargo/util/toml_mut/upgrade.rs [hunk 1]
```diff
@@ -217,5 +223,10 @@ mod test {
             assert_req_bump("1.1.1", "=1.0.0", "=1.1.1");
             assert_req_bump("2.0.0", "=1.0.0", "=2.0.0");
         }
+
+        #[test]
+        fn caret_prerelease() {
+            assert_req_bump("1.7.0", "2.0.0-beta.21", None);
```
Comment (weihanglo) [inline src/cargo/util/toml_mut/upgrade.rs:229]: Could we test against other SemVer operators? Like `assert_req_bump("1.7.0", ">=2.0.0-beta.21", None)`. We might want to ignore this one instead of panicking. Thought?

File: src/cargo/util/toml_mut/upgrade.rs [hunk 2]
```diff
@@ -77,7 +80,10 @@ fn set_comparator(
 fn assign_partial_req(
     version: &semver::Version,
     mut pred: semver::Comparator,
-) -> semver::Comparator {
+) -> Option<semver::Comparator> {
+    if !pred.pre.is_empty() {
```
Comment (weihanglo) [inline src/cargo/util/toml_mut/upgrade.rs:84]: Better to have common for the meaning of `Option` here or on `set_comparator`.

**Comment by linyihai** [comment]
wait a second, I squashed the two commit into one accidentally, let me split them.

File: src/cargo/util/toml_mut/upgrade.rs (hunk 1 above)
Comment (linyihai) [reply src/cargo/util/toml_mut/upgrade.rs:232]: Ok, I had added other Semver operators test but in the renamed test `ignore_prerelease`

File: src/cargo/util/toml_mut/upgrade.rs (hunk 2 above)
Comment (linyihai) [reply src/cargo/util/toml_mut/upgrade.rs]: That's a good idea. I moved the changes up to the `upgrade_requirement` to avoid fixing them here

**Comment by linyihai** [comment]
```
---- expected: tests/testsuite/https.rs:159:27
++++ actual:   stderr
   1    1 | [UPDATING] git repository `[https://github.com/rust-lang/bitflags.git`](https://github.com/rust-lang/bitflags.git%60)
        2 + [WARNING] spurious network error (3 tries remaining): failed to connect to github.com: Connection timed out; class=Os (2)
   2    3 | [LOCKING] 2 packages to latest compatible versions

```

May be supcious network connection error?

File: src/cargo/util/toml_mut/upgrade.rs (hunk 1 above)
Comment (weihanglo) [reply src/cargo/util/toml_mut/upgrade.rs:232]: My fault. I think I have missed this <https://github.com/rust-lang/cargo/issues/14178#issuecomment-2203083535>. Let's try clarifying it :)

@epage Did you suggest allowing prerelease upgrade under the current stable semver semantic? Should we instead ignore prerelease and document the behavior, until we settle the semantic in `cargo update --precise <prerelease>` feature?

Comment (linyihai) [reply src/cargo/util/toml_mut/upgrade.rs:232]: https://github.com/rust-lang/cargo/blob/a2b58c3dad4d554ba01ed6c45c41ff85390560f2/src/cargo/ops/cargo_update.rs#L330-L336

The `update --breaking` had filtered the `prerelease` version to update to. So here equals to no upgrade to prerelease also. :/

**Comment by epage** [comment]
CC @torhovland

I feel like one of your PRs had touched on this already but we wanted to split it out for focusing the reviews,  or am I thinking of something else?

**Comment by torhovland** [comment]
> CC @torhovland
> 
> I feel like one of your PRs had touched on this already but we wanted to split it out for focusing the reviews, or am I thinking of something else?

I think these are separate issues. This one here is about `update --breaking` not respecting that the user already has a higher pre-release installed, thus trying to "upgrade" to a lower version.

The one I have encountered is about a precise version req not matching its exact version if it is a pre-release. See https://github.com/rust-lang/cargo/pull/14140/commits/de1a4a2512cac7d6e8e698fc0f3637893543d708

**Comment by linyihai** [comment]
I copied the  `matches_greater` from [`semver`](https://github.com/dtolnay/semver/blob/master/src/eval.rs#L64) to make sure prerelease VersionReq won't downgrade.

File: tests/testsuite/update.rs [hunk 3]
```diff
@@ -2616,3 +2616,37 @@ fn update_breaking_mixed_pinning_renaming() {
...
+        [dependencies]
+        bar = "2.0.0-beta.21"
+    "#,
+        )
+        .file("src/lib.rs", "")
+        .build();
+
+    p.cargo("generate-lockfile").run();
+
+    Package::new("bar", "1.7.0").publish();
```
Comment (torhovland) [inline tests/testsuite/update.rs:2643]: A comment would be useful here.

```
// The purpose of this test is to demonstrate that `update --breaking` will not try to downgrade to the latest stable version (1.7.0), but will rather keep the latest pre-release (2.0.0-beta.21).
```

Comment (linyihai) [reply tests/testsuite/update.rs:2646]: Good idea. Have already added.

File: src/cargo/util/toml_mut/upgrade.rs [hunk 4]
```diff
@@ -18,9 +18,18 @@ pub(crate) fn upgrade_requirement(
...
             .comparators
             .into_iter()
-            .map(|p| set_comparator(p, version))
+            .filter_map(|p| {
+                if p.pre.is_empty() {
+                    Some(set_comparator(p, version))
+                } else {
+                    None
+                }
+            })
```
[the remark this answers is not in the document]
Comment (linyihai) [reply src/cargo/util/toml_mut/upgrade.rs]: Done, thanks

File: tests/testsuite/update.rs [hunk 5]
```diff
@@ -2616,3 +2616,39 @@ fn update_breaking_mixed_pinning_renaming() {
             "#]],
     );
 }
+
+#[cargo_test]
+fn update_breaking_pre_release() {
```
Comment (weihanglo) [inline tests/testsuite/update.rs:2621]: From <https://github.com/rust-lang/cargo/pull/14250/files#r1680348122>:

> The `update --breaking` had filtered the `prerelease` version to update to. So here equals to no upgrade to prerelease also. :/

Since that is the case, we might want to have UI tests for

1. `2.0.0-beta.22` is available, but `2.0.0-beta.21` won't upgrade to it.
2. `2.0.0` is available, but `2.0.0-beta.21` won't upgrade to it.
2. `3.0.0` is available, and `2.0.0-beta.21` will upgrade to it.

We should document the first two as an unresolved issue. That is something we want to support, especially the second one.

Comment (linyihai) [reply tests/testsuite/update.rs]: Added `update_breaking_pre_release_upgrade` to reflect the current behavior.

> 2.0.0 is available, but 2.0.0-beta.21 won't upgrade to it.

This case out of my expection, it seems need to support.

**Review by weihanglo — APPROVED** [review]

**Comment by weihanglo** [comment]
While there are still some edge cases having to be supported, we shouldn't block this from fixing wrong downgrades. Thank you for the contribution!

@bors r+

File: tests/testsuite/update.rs (hunk 5 above)
Comment (weihanglo) [reply tests/testsuite/update.rs]: Task added: <https://github.com/rust-lang/cargo/issues/12425#issuecomment-2186198258>.
